### PR TITLE
[Planning]  Notebooks

### PR DESCRIPTION
**Pull request by lucabergamini**
PR for planning notebooks

### CODE REVIEW

File: examples/drivenet/drivenet_train.ipynb [hunk 1]
```diff
@@ -0,0 +1,325 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Introduction to DriveNet\n",
+    "\n",
+    "DriveNet is a CNN able to drive an agent on a specific route. Among its features, DriveNet can follow the midlane and respect traffic lights.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:9]: Nit: I'd actually use 'DNN'

Among others, DriveNet can do lane-following while respecting traffic lights.

File: examples/drivenet/drivenet_train.ipynb [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "### Imitation Learning and Its limitations\n",
+    "DriveNet is trained using imitation learning on our Lyft L5 Prediction Dataset 2020. We feed examples of real driving experience to the model and we expect it to take the same actions as the driver did in those episodes. This is very similar to how tasks like classification are usually solved.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:12]: using the principle of ..., tasked with learning from expert recordings on our ...

Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:12]: experienceS, remove we in "and we expect".
Maybe explicitly mention supervised learning?

File: examples/drivenet/drivenet_train.ipynb [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "Imitiation Learning is powerful, but it has a strong limitation. Even with a lot of data, is easy to drift away from the original train distribution. This may not be an issue for single-shot task (like classification), but this limit becomes immediately clear for autoregressive predictions.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:14]: IT is easy.

I'd slightly reformulate though and also use terms compounding errors and distriibution or similar.
Mention that model has only see states along expert samples, and that errors accumulate over time while unrolling which becomes a problem.

File: examples/drivenet/drivenet_train.ipynb [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "After training DriveNet, we would like it to take full control of the AV and determine its future locations. This means that subsequent predictions will be use as future locations for the AV. \n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:16]: useD.

Maybe this is too much in our 'language' though? At this point, readers might not think of modifying future positions in a zarr etc -> just say DN controls AV or so?

File: examples/drivenet/drivenet_train.ipynb [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "In this scenario it's very easy to drift away from the original distribution in the train set. During training our DriveNet has seen good examples of driving only. In particular, this means **almost perfect midlane following**. However, even a small constant vertical displacement during the close loop test can accumulate enough error to lead the AV completely out of its distribution.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:18]: Kind of a repetition ...

Comment (oliver-scheel) [reply examples/drivenet/drivenet_train.ipynb:18]: Would reformulate with above ... does vertical drift has to be constant?

Comment (oliver-scheel) [reply examples/drivenet/drivenet_train.ipynb:18]: I think it's CLOSED-LOOP

File: examples/drivenet/drivenet_train.ipynb [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "In this notebook you will see an example of how we can tackle this issue effectively and learn to recover from out-of-distribution situations.\n",
+    "\n",
+    "TODO PICTURE OF WHAT DRIFTING MEANS HERE\n",
+    "\n",
+    "\n",
+    "In this notebook you're going to train your own Drivenet model using Lyft L5 Dataset and L5Kit.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:25]: the ... Dataset - it's called differently though, isn't it?

File: examples/drivenet/drivenet_train.ipynb [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "**Before starting, please download the [Lyft L5 Prediction Dataset 2020](https://self-driving.lyft.com/level5/prediction/) and follow [the instructions](https://github.com/lyft/l5kit#download-the-datasets) to correctly organise it**\n"
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:27]: .

File: examples/drivenet/drivenet_train.ipynb [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    "cfg = load_config_data(\"./drivenet_config.yaml\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Adding Perturbations to the mix\n",
+    "\n",
+    "One of the simpler techniques to ensure a good generalisation is **data augmentation**, which exposes the network to different version of the input and helps it to generalise better to out-of-distribution situations.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:87]: versions

File: examples/drivenet/drivenet_train.ipynb [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    "# before leaving, ensure perturb_prob is correct\n",
+    "perturbation.perturb_prob = perturb_prob\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model\n",
+    "L5Kit provides a model file for DriveNet. The backbone is a ResNetX one (either 18 or 50) pre-trained on ImageNet. Here, we describe the main inputs and outputs of the model; for a full description please check `l5kit/drivenet/model.py`.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:142]: Backbone is ResNetX (either ...) ..

Comment (oliver-scheel) [reply examples/drivenet/drivenet_train.ipynb:142]: Link to file

File: examples/drivenet/drivenet_train.ipynb [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "#### Inputs\n",
+    "L5Kit is shipped with various rasterisers. Each one capture the information around the AV and projects it into a fixed grid of pixels we can feed to our CNN backbone. Each rasteriser has its own input representation and in general it's not an RGB image. As an example, the ego bounding box and agents are stored in additional channels in the semantic rasteriser.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:145]: , before general

File: examples/drivenet/drivenet_train.ipynb [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    "TODO IMAGES\n",
+    "\n",
+    "#### Outputs\n",
+    "\n",
+    "The model outputs change between train an evaluation. During train, the loss value is computed and returned, while the full outputs are reported during the evaluation phase. These are the future positional and angular offsets."
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:151]: and

Comment (oliver-scheel) [reply examples/drivenet/drivenet_train.ipynb:151]: I would rephrase this, the model output's dont change (mathematically).

File: examples/drivenet/drivenet_train.ipynb [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    "        )\n",
+    "print(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Prepare for training\n",
+    "Our `EgoDataset` inherits from pytorch `Dataset`, so we can use it inside a `Dataloader` to enable multi-processing"
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:175]: Pytorch

File: examples/drivenet/drivenet_train.ipynb [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+   "metadata": {},
+   "source": [
+    "# What's Next\n",
+    "\n",
+    "**Congratulations in training your first DriveNet model!**\n",
+    "\n",
+    "\n",
+    "Now that your model is trained and safely stored, you can evaluate how it performs in different situations.\n",
+    "\n",
+    "First, you can check if it can produce sensible predictions in the easier **open loop** settings using the dedicated notebook (TODO LINK). \n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:289]: setting

File: examples/drivenet/drivenet_train.ipynb (hunk 6 above)
Comment (lucabergamini) [reply examples/drivenet/drivenet_train.ipynb:25]: I can remove prediction, but that's the name of the dataset in the end xD

File: examples/drivenet/drivenet_config.yaml [hunk 14]
```diff
@@ -0,0 +1,66 @@
+# Config format schema number
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_config.yaml:1]: Would it make sense to have a common config folder?
Now we have at least 3 (same) configs across the examples - or is that badly overloaded due to model_params being once for SimNet and once for DriveNet?

File: examples/drivenet/drivenet_train.ipynb [hunk 15]
```diff
@@ -0,0 +1,321 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Introduction to DriveNet\n",
+    "\n",
+    "DriveNet is a DNN able to drive an agent on a specific route. Among others, DriveNet can do lane-following while respecting traffic lights.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:9]: along a specific

File: examples/drivenet/drivenet_train.ipynb [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+    "\n",
+    "### Imitation Learning and Its limitations\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:11]: Inconsistent casing - do you capitalize all words in titles or just the first? (seems inconsistent across headings).
If you capitalize 'all' -> Limitations, but don't capitalize fill words like its

File: examples/drivenet/drivenet_train.ipynb [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+    "DriveNet is trained using imitation learning on our Lyft L5 Prediction Dataset 2020. We feed examples of real driving experiences to the model and expect it to take the same actions as the driver did in those episodes. This is very similar to how tasks like classification are usually solved.\n",
+    "\n",
+    "Imitiation Learning is powerful, but it has a strong limitation. It's not trivial for a trained model to generalise well on out-of-distribution data.\n",
+    "\n",
+    "After training DriveNet, we would like it to take full control and drive the AV in an autoregressive fashion (i.e. by following its own predictions).\n",
+    "\n",
+    "In this scenario it's very easy for errors to compound and make the AV drift away from the original distribution. During training our DriveNet has seen good examples of driving only. In particular, this means **almost perfect midlane following**. However, even a small constant displacement during evaluate can accumulate enough error to lead the AV completely out of its distribution in a matter of seconds.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:18]: during evaluation

File: examples/drivenet/drivenet_train.ipynb [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+    "\n",
+    "![drifting](../../images/drivenet/drifting.svg)\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:20]: VERY NICE!
What is the little top part coming in though, is it a lane?
It looks too narrow for a lane and is confusing.

File: examples/drivenet/drivenet_train.ipynb [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+    "\n",
+    "In this notebook you're going to train your own Drivenet model using Lyft L5 Dataset and L5Kit. You will use a technique named **online trajectory perturbation** to mitigate the effects of drifting.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:22]: THE Lyft L5 ... it also has prediction in the name doesn't it?

Comment (oliver-scheel) [reply examples/drivenet/drivenet_train.ipynb:22]: called instead of named

Comment (oliver-scheel) [reply examples/drivenet/drivenet_train.ipynb:22]: Maybe explain what in the code 'Ackermann' in perturbation is?

File: examples/drivenet/drivenet_train.ipynb [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+    "    perturbation.perturb_prob = perturbation_value\n",
+    "\n",
+    "    data_ego = train_dataset[0]\n",
+    "    im_ego = rasterizer.to_rgb(data_ego[\"image\"].transpose(1, 2, 0))\n",
+    "    target_positions = transform_points(data_ego[\"target_positions\"], data_ego[\"raster_from_agent\"])\n",
+    "    draw_trajectory(im_ego, target_positions, TARGET_POINTS_COLOR)\n",
+    "    plt.imshow(im_ego[::-1])\n",
+    "    plt.axis('off')\n",
+    "    plt.show()\n",
+    "    \n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:127]: too man blank lines

File: examples/drivenet/drivenet_train.ipynb [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+    "# before leaving, ensure perturb_prob is correct\n",
+    "perturbation.perturb_prob = perturb_prob\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model\n",
+    "L5Kit provides a model file for DriveNet. The backbone is a ResNetX(either 18 or 50) model pre-trained on ImageNet. Here, we describe the main inputs and outputs of the model; for a full description please check [the class definition](https://github.com/lyft/l5kit/blob/1d077fdbc8656057516bee2bb6cc815bc6868d29/l5kit/l5kit/drivenet/model.py#L9).\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:139]: space after X

File: examples/drivenet/drivenet_train.ipynb [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+    "\n",
+    "#### Inputs\n",
+    "L5Kit is shipped with various rasterisers. Each one capture the information around the AV and projects it into a fixed grid of pixels we can feed to our CNN backbone. Each rasteriser has its own input representation and, in general, it's not an RGB image. As an example, the ego bounding box and agents are stored in additional channels in the semantic rasteriser.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:142]: captures

File: examples/drivenet/drivenet_train.ipynb [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+    "        )\n",
+    "print(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Prepare for training\n",
+    "Our `EgoDataset` inherits from PyTorch `Dataset`, so we can use it inside a `Dataloader` to enable multi-processing"
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:173]: . or :

File: examples/drivenet/drivenet_config.yaml (hunk 14 above)
Comment (lucabergamini) [reply examples/drivenet/drivenet_config.yaml:1]: they are slightly different so I think we should leave them in different folder for now. Definitely something to consider for later though

File: examples/drivenet/drivenet_train.ipynb (hunk 19 above)
Comment (lucabergamini) [reply examples/drivenet/drivenet_train.ipynb:22]: You're probably the best one who can do that lol

File: examples/drivenet/drivenet_train.ipynb [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+   "metadata": {},
+   "source": [
+    "# Introduction to DriveNet\n",
+    "\n",
+    "DriveNet is a DNN able to drive an agent along a specific route. Among others, DriveNet can do lane-following while respecting traffic lights.\n",
+    "\n",
+    "### Imitation learning and its limitations\n",
+    "DriveNet is trained using imitation learning on our Lyft L5 Prediction Dataset 2020. We feed examples of real driving experiences to the model and expect it to take the same actions as the driver did in those episodes. This is very similar to how tasks like classification are usually solved.\n",
+    "\n",
+    "Imitiation Learning is powerful, but it has a strong limitation. It's not trivial for a trained model to generalise well on out-of-distribution data.\n",
```
Comment (melights) [inline examples/drivenet/drivenet_train.ipynb:14]:
```suggestion
    "Imitation Learning is powerful, but it has a strong limitation. It's not trivial for a trained model to generalise well on out-of-distribution data.\n",
```

File: examples/drivenet/drivenet_train.ipynb [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+    "## Model\n",
+    "L5Kit provides a model file for DriveNet. The backbone is a ResNetX (either 18 or 50) model pre-trained on ImageNet. Here, we describe the main inputs and outputs of the model; for a full description please check [the class definition](https://github.com/lyft/l5kit/blob/1d077fdbc8656057516bee2bb6cc815bc6868d29/l5kit/l5kit/drivenet/model.py#L9).\n",
+    "\n",
+    "#### Inputs\n",
+    "L5Kit is shipped with various rasterisers. They all capture the information around the AV and project it into a fixed grid of pixels we can feed to our CNN backbone. Each rasteriser has its own input representation and, in general, it's not an RGB image. As an example, the ego bounding box and agents are stored in additional channels in the semantic rasteriser.\n",
+    "\n",
+    "\n",
+    "#### Outputs\n",
+    "\n",
+    "During train, the loss value is computed and returned, while the full outputs are returned during the evaluation phase. These are the future positional and angular offsets.\n",
```
Comment (melights) [inline examples/drivenet/drivenet_train.ipynb:145]:
```suggestion
    "During training, the loss value is computed and returned, while the full outputs are returned during the evaluation phase. These are the future positional and angular offsets.\n",
```

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 26]
```diff
@@ -0,0 +1,354 @@
...
+    "print(eval_dataset)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Evaluation loop\n",
+    "\n",
+    "Here, we loop through the data and store predicted and annotated trajectories (postions + yaws)\n",
```
Comment (melights) [inline examples/drivenet/drivenet_open_loop.ipynb:126]:
```suggestion
    "Here, we loop through the data and store predicted and annotated trajectories (positions + yaws)\n",
```

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 27]
```diff
@@ -0,0 +1,354 @@
...
+    "\n",
+    "#### Positional Displacement\n",
+    "[Average Displacement Error (ADE) and Final Displacement Error(FDE)](https://en.wikipedia.org/wiki/Mean_squared_displacement) are standard metrics used to evaluate future predictions for AVs.\n",
+    "\n",
+    "We can compute them by comparing predicted and annotated positions, which we have stored in the previous cell.\n",
+    "Additionally, we can plot histograms of their distributions across samples to better capture the variance of our error\n",
+    "\n",
+    "#### Angle Displacement\n",
+    "\n",
+    "For the yaw, we can use the Minimum Angle Distance to check the error. Again, we can plot an histogram to inspect the error distribution."
```
Comment (melights) [inline examples/drivenet/drivenet_open_loop.ipynb:180]:
```suggestion
    "For the yaw, we can use the Minimum Angle Distance to check the error. Again, we can plot a histogram to inspect the error distribution."
```

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 28]
```diff
@@ -0,0 +1,354 @@
...
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Qualitative Evaluation: Visualise the Open Loop\n",
+    "\n",
+    "To visualise the open loop we can just repeat the same operations for consecutives frames.\n",
```
Comment (melights) [inline examples/drivenet/drivenet_open_loop.ipynb:268]:
```suggestion
    "To visualise the open loop we can just repeat the same operations for the consecutive frames.\n",
```

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 29]
```diff
@@ -0,0 +1,354 @@
...
+   "source": [
+    "## Is open loop evaluation enough?\n",
+    "\n",
+    "Depending on how long you've trained your model, you may have seen different results in the above cell.\n",
+    "\n",
+    "If you've trained it long enough, the predicted trajectory is likely well overlapped with the annotated one.\n",
+    "\n",
+    "Conversely, a model trained not enough will surely show some bias or unfeasible trajectories.\n",
+    "\n",
+    "In both cases, **this evaluation is not enough** to ensure your model will be able to actually drive on the road (that's where we all want to go in the end). If your model is not in controll of where to go, you can't really say it will work once the annotated trajectory won't be available anymore.\n",
```
Comment (melights) [inline examples/drivenet/drivenet_open_loop.ipynb:318]:
```suggestion
    "In both cases, **this evaluation is not enough** to ensure your model will be able to actually drive on the road (that's where we all want to go in the end). If your model is not in control of where to go, you can't really say it will work once the annotated trajectory won't be available anymore.\n",
```

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 30]
```diff
@@ -0,0 +1,329 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# DriveNet Closed-Loop Evaluation\n",
+    "\n",
+    "**Note: this notebook assumes you've already run the [training notebook](./drivenet_train.ipynb) and stored your model successfully.**\n",
+    "\n",
+    "## What is a closed-loop evaluation?\n",
+    "In closed-loop evaluation DriveNet is in **full control of the AV**. At each time step, we predict the future trajectory and then move the AV int he first of the DriveNet's predictions. \n",
```
Comment (melights) [inline examples/drivenet/drivenet_close_loop.ipynb:12]:
```suggestion
    "In closed-loop evaluation DriveNet is in **full control of the AV**. At each time step, we predict the future trajectory and then move the AV into the first of the DriveNet's predictions. \n",
```

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+    "\n",
+    "## What is a good closed-loop metric?\n",
+    "In this setting **metrics are particularly challenging**. In fact, we would like to penalise some of the simulation drift (e.g. going off road or in the opposite lane) while at the same time allow others (e.g. different speed profiles). As an example, using the same metrics we employed during the open loop evaluation (e.g. ADE) would penalise DriveNet Unconditionally.\n",
+    "\n",
+    "At Lyft L5, we use a large set of different metrics to capture dangerous manoeuvres and behaviours. \n",
+    "\n",
+    "For the sake of simplicity, in this notebook we will be using a very simple proxy to detect if our model is driving in a sensible way, composed of two different metrics:\n",
+    "\n",
+    "### Collisions\n",
+    "Our AV should avoid collisions with other agents. However, not all collisions are created equally. This is because while our AV is fully controlled by Drivenet, other agents are not (a setting we call **log replay**).\n",
```
Comment (melights) [inline examples/drivenet/drivenet_close_loop.ipynb:29]:
```suggestion
    "Our AV should avoid collisions with other agents. However, not all collisions are created equally. This is because while our AV is fully controlled by DriveNet, other agents are not (a setting we call **log replay**).\n",
```

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# DriveNet Closed-Loop Evaluation\n",
+    "\n",
+    "**Note: this notebook assumes you've already run the [training notebook](./drivenet_train.ipynb) and stored your model successfully.**\n",
+    "\n",
+    "## What is a closed-loop evaluation?\n",
+    "In closed-loop evaluation DriveNet is in **full control of the AV**. At each time step, we predict the future trajectory and then move the AV into the first of the DriveNet's predictions. \n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:12]: move TO the first ..

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 33]
```diff
@@ -0,0 +1,329 @@
...
+    "\n",
+    "At Lyft, we refer to this process with the terms **forward-simulate** or **unroll**.\n",
+    "\n",
+    "![closed-loop](../../images/drivenet/closed-loop.svg)\n",
+    "\n",
+    "This evalaution is crucial to assess the real performance of our DriveNet. **Ideally, we would test the model on the road in the real world**. However, this is clearly very expensive and scales poorly. Forward-simulation is an attempt to evaluate the system in a setting which is closest as possible to a real road test on the same route.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:18]: as close as possible

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 34]
```diff
@@ -0,0 +1,329 @@
...
+    "\n",
+    "\n",
+    "## What is a good closed-loop metric?\n",
+    "In this setting **metrics are particularly challenging**. In fact, we would like to penalise some of the simulation drift (e.g. going off road or in the opposite lane) while at the same time allow others (e.g. different speed profiles). As an example, using the same metrics we employed during the open loop evaluation (e.g. ADE) would penalise DriveNet Unconditionally.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:22]: unconditionally

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 35]
```diff
@@ -0,0 +1,329 @@
...
+    "At Lyft L5, we use a large set of different metrics to capture dangerous manoeuvres and behaviours. \n",
+    "\n",
+    "For the sake of simplicity, in this notebook we will be using a very simple proxy to detect if our model is driving in a sensible way, composed of two different metrics:\n",
+    "\n",
+    "### Collisions\n",
+    "Our AV should avoid collisions with other agents. However, not all collisions are created equally. This is because while our AV is fully controlled by DriveNet, other agents are not (a setting we call **log replay**).\n",
+    "\n",
+    "If our AV was slower than the recorded one, chasing agents might bump into us. Clearly, this won't happen in a real setting where other agents can react to our behaviour.\n",
+    "\n",
+    "In this simple example, we won't distinguish between collisions caused by non reactivity of other agents and actual collisions, and we will simply report them all categorised by where they occurred (front, rear and side with respect to the AV).\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:33]: non-reactivity

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 36]
```diff
@@ -0,0 +1,329 @@
...
+    "\n",
+    "However, if we only considered collision, our AV could pass our tests by driving off-road or in a different lane.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:35]: collisions.
Repetition of ours, woud suggest all tests

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 37]
```diff
@@ -0,0 +1,329 @@
...
+    "\n",
+    "### Distance from Reference Trajectory\n",
+    "To address the issue presented above, we require our DriveNet to loosely stick to the original trajectory in the data. By setting the right threshold on the distance we can allow for different speed profile and small steerings, while pensalising large deviations like driving off-road.\n",
+    "\n",
+    "We can do so by computing the distance between the first predictions from DriveNet and the corresponding annotated positions in world coordinates."
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:44]: Could wee have a heading here or something?
Looks like this cell is about dist. to ref. traj.

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 38]
```diff
@@ -0,0 +1,354 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# DriveNet Open Loop Evaluation\n",
+    "\n",
+    "**Note: this notebook assumes you've already run the [training notebook](./drivenet_train.ipynb) and stored your model successfully.**\n",
+    "\n",
+    "### What is open loop evaluation?\n",
+    "In open-loop evaluation we evaluate our model prediction as we follow the annotated ground truth.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_open_loop.ipynb:12]: I'm actually not sure now whether its open loop or open-loop - but could u make it consistent?

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 39]
```diff
@@ -0,0 +1,354 @@
...
+    "\n",
+    "In each frame, we compare the predictions of our model against the annotated ones. This can be done with different metrics, and we will see few of them in the following.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_open_loop.ipynb:14]: annotates ones -> mh it's not annotated predictions, GT?

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 40]
```diff
@@ -0,0 +1,354 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load the Evaluation Data\n",
+    "This is almost the exact same code you've already seen in the [training notebook](./drivenet_train.ipynb). Apart from the different dataset we load, the biggest difference is that **we don't perturb our data here**.\n",
+    "\n",
+    "When performing evaluation we're interested in knowing the performance on the annotated data, not on perturbed ones."
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_open_loop.ipynb:101]: one

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 41]
```diff
@@ -0,0 +1,354 @@
...
+    "print(eval_dataset)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Evaluation loop\n",
+    "\n",
+    "Here, we loop through the data and store predicted and annotated trajectories (positions + yaws)\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_open_loop.ipynb:126]: .

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 42]
```diff
@@ -0,0 +1,354 @@
...
+    "\n",
+    "Note: we're not taking into account availability here. We acknowledge this can reflects in lower metric."
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_open_loop.ipynb:128]: What does this mean?
Also: reflect

File: examples/drivenet/drivenet_open_loop.ipynb [hunk 43]
```diff
@@ -0,0 +1,354 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Quantitative Evaluation: ADE, FDE and Angle Distance\n",
+    "\n",
+    "#### Positional Displacement\n",
+    "[Average Displacement Error (ADE) and Final Displacement Error(FDE)](https://en.wikipedia.org/wiki/Mean_squared_displacement) are standard metrics used to evaluate future predictions for AVs.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_open_loop.ipynb:173]: space before (FDE)

File: examples/drivenet/drivenet_train.ipynb [hunk 44]
```diff
@@ -0,0 +1,321 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Introduction to DriveNet\n",
+    "\n",
+    "DriveNet is a DNN able to drive an agent along a specific route. Among others, DriveNet can do lane-following while respecting traffic lights.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:9]: Just a though: could extend this a bit? Maybe one or two more sentences what we want to do (drive the car, using data-driven methods compared to expert planner, use largs amounts of data etc ... )

File: examples/drivenet/drivenet_train.ipynb [hunk 45]
```diff
@@ -0,0 +1,321 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model\n",
+    "L5Kit provides a model file for DriveNet. The backbone is a ResNetX (either 18 or 50) model pre-trained on ImageNet. Here, we describe the main inputs and outputs of the model; for a full description please check [the class definition](https://github.com/lyft/l5kit/blob/1d077fdbc8656057516bee2bb6cc815bc6868d29/l5kit/l5kit/drivenet/model.py#L9).\n",
+    "\n",
+    "#### Inputs\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_train.ipynb:139]: Maybe link / mention agent prediction here, mention that inputs etc are very similar?

**Review by oliver-scheel — COMMENTED** [review]
Amazing notebooks!

File: examples/drivenet/drivenet_open_loop.ipynb (hunk 42 above)
Comment (lucabergamini) [reply examples/drivenet/drivenet_open_loop.ipynb:128]: I'm not storing availabilities and use them during metric computation to filter out some elements. It's not a big deal, I'll just remove the sentence and we can address it later if we want to
